File: Binary_Tree/543.Diameter_of_Binary_Tree/C++/543.Diameter_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode * root) 
    {
        //==========================================//
        // Input type:                              //
        //   - TreeNode * (root tree-node)          //
        // Return type:                             //
        //   - int (result maximun summary counter) //
        //==========================================//
        if (root && ((! (root->left)) && (! (root->right)))) { return 0; } //Issue, Boundary-case handling

        /*Initialize*/
        ///// Result maximun summary counter /////
        int res_max_cnter = 0;

        /*Recursion-based DFS loop traversal*/
        int res_val = dfsTraversal(root, res_max_cnter); //Recursion function call (whole root node)

        return res_max_cnter;
    }


    int dfsTraversal(TreeNode * root, int & res_max_cnter)
    {
        //===============================================//
        // Input type:                                   //
        //   - TreeNode * (root tree-node)               //
        //   - int ref. (result maximun summary counter) //
        // Return type:                                  //
        //   - int (result maximun summary counter)      //
        //===============================================//
        if (! root) { return 0; } //Issue, Boundary-case handling

        if ((! root->left) && (! (root->right))) { return 1; } //Issue, Boundary-case handling


        /*Whole process, flow*/
        int record_left_cnter = dfsTraversal((root->left), res_max_cnter); //Recursion function call (root left-half node)
        int record_right_cnter = dfsTraversal((root->right), res_max_cnter); //Recursion function call (root right-half node)

        res_max_cnter = max(res_max_cnter, (record_left_cnter + record_right_cnter)); //Keep updating, overwriting

        return (1 + max(record_left_cnter, record_right_cnter));
    }
};

File: Binary_Tree/543.Diameter_of_Binary_Tree/Python/543.Diameter_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#
        if (root and (not root.left) and (not root.right)): #Issue/Boundary-case handling
            return 0

        ############
        #Initialize
        ##### Result maximun summary counter ######
        res_max_cnter = [0]

        ####################################
        #Recursion-based DFS loop traversal
        self.dfsTraversal(root, res_max_cnter) #Recursion function call

        return res_max_cnter[0]


    def dfsTraversal(self, root, res_max_cnter):
        """
        :type root: Optional[TreeNode]
        :type res_max_cnter: List[int]
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (not root): #Issue/Boundary-case handling
            return 0
        
        if ((not root.left) and (not root.right)): #Issue/Boundary-case handling
            return 1

        ####################
        #Whole process/flow
        record_left_cnter = self.dfsTraversal(root.left, res_max_cnter) #Recursion function call (i.e. root left-half node)
        record_right_cnter = self.dfsTraversal(root.right, res_max_cnter) #Recursion function call (i.e. root right-half node)

        res_max_cnter[0] = max(res_max_cnter[0], (record_left_cnter + record_right_cnter)) #Keep updating/overwriting

        return (max(record_left_cnter, record_right_cnter) + 1)
